### PR TITLE
Feature/aql optimizations

### PR DESCRIPTION
**Pull request by jsteemann**

### CODE REVIEW

File: arangod/Aql/Ast.cpp [hunk 1]
```diff
@@ -2217,125 +2217,54 @@ std::unordered_set<std::string> Ast::getReferencedAttributesForKeep(AstNode cons
...
-bool Ast::variableOnlyUsedForSingleAttributeAccess(AstNode const* node,
-                                                   Variable const* variable,
-                                                   std::vector<std::string> const& attributeName) {
-  bool result = true;
-
+/// @brief determines the top-level attributes referenced in an expression for the
+/// specified out variable
+bool Ast::getReferencedAttributes(AstNode const* node, 
+                                  Variable const* variable,
+                                  std::unordered_set<std::string>& vars) {
```
Comment (mchacki) [inline arangod/Aql/Ast.cpp]: Is there a reason why this cannot be an attributePath?
e.g.:
`a.b.c` would not be optimized t was before

File: arangod/Aql/ClusterNodes.cpp [hunk 2]
```diff
@@ -131,11 +131,20 @@ DistributeNode::DistributeNode(ExecutionPlan* plan,
...
+      _variable(nullptr),
+      _alternativeVariable(nullptr),
       _createKeys(base.get("createKeys").getBoolean()),
       _allowKeyConversionToObject(base.get("allowKeyConversionToObject").getBoolean()),
-      _allowSpecifiedKeys(false) {}
+      _allowSpecifiedKeys(false) {
+ 
+  if (base.hasKey("variable")) {     
+    _variable = Variable::varFromVPack(plan->getAst(), base, "variable");
+    _alternativeVariable = Variable::varFromVPack(plan->getAst(), base, "alternativeVariable");
```
Comment (mchacki) [inline arangod/Aql/ClusterNodes.cpp]: what if base does not have alternativeVariable?
i would say the test should be:
`base.hasKey("variable") && base.hasKey("alternativeVariable")`
or two ifs if they can work without one another.

File: arangod/Aql/IndexNode.cpp [hunk 3]
```diff
@@ -158,6 +224,10 @@ ExecutionNode* IndexNode::clone(ExecutionPlan* plan, bool withDependencies,
 
   auto c = new IndexNode(plan, _id, _vocbase, _collection, outVariable,
                          _indexes, _condition->clone(), _reverse);
+  
+  c->projections(_projections);
+  c->needsGatherNodeSort(_needsGatherNodeSort);
```
Comment (mchacki) [inline arangod/Aql/IndexNode.cpp]: This line is now twice

File: arangod/RocksDBEngine/RocksDBEdgeIndex.h [hunk 4]
```diff
@@ -57,10 +57,16 @@ class RocksDBEdgeIndexIterator final : public IndexIterator {
...
+  bool nextCovering(DocumentCallback const& cb, size_t limit) override;
   bool nextExtra(ExtraCallback const& cb, size_t limit) override;
   void reset() override;
+  
+  /// @brief we provide a method to provide the index attribute values
+  /// while scanning the index
+  bool hasCovering() const override { return true; }
 
  private:
+  bool initKey(arangodb::velocypack::Slice& key);
```
Comment (mchacki) [inline arangod/RocksDBEngine/RocksDBEdgeIndex.h]: Please add a comment that caller needs to make sure that the Slice data is not deleted as long as this Iterator is ongoing.
The Iterator will only retain the Slice and will not copy out the data.
If this is forgotten we end up in Undefined behaviour.

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr-linux/393/

**Pull request merged**